### PR TITLE
Clarify test features in product code

### LINKED ISSUE #247: Misleading phrasing: changing code to make it testable
Hi all,

I stumbled across this phrasing:

> ["Don't change the productive code to make the code testable"](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#dont-change-the-productive-code-to-make-the-code-testable)

The intended meaning is clarified by the example, but the heading is really misleading. There is nothing wrong with changing productive code in order to make it testable. Mind that "testable code" is not "test code". Testable code is always a major design goal and we often refactor towards this goal. The issue is when productive code contains test code. I would suggest another phrasing, for example:

- "Don't let test code invade producitve code" or
- "Productive code must not contain test code"

Best,
Jonathan

### PR DESCRIPTION
**Pull request by pokrakam**
Closes #247

### CODE REVIEW

File: clean-abap/CleanABAP.md [hunk 1]
```diff
@@ -219,7 +219,7 @@ The [Cheat Sheet](cheat-sheet/CheatSheet.md) is a print-optimized version.
     - [Use test seams as temporary workaround](#use-test-seams-as-temporary-workaround)
     - [Use LOCAL FRIENDS to access the dependency-inverting constructor](#use-local-friends-to-access-the-dependency-inverting-constructor)
     - [Don't misuse LOCAL FRIENDS to invade the tested code](#dont-misuse-local-friends-to-invade-the-tested-code)
-    - [Don't change the productive code to make the code testable](#dont-change-the-productive-code-to-make-the-code-testable)
+    - [Don't add unit test features to productive code](#dont-add-unit-test-features-to-productive-code)
```
[the remark this answers is not in the document]
Comment (N2oB6n-SAP) [reply clean-abap/CleanABAP.md:222]: I agree about the misleading title. The arguments in #247 nicely highlight that productive code needs to be changed to be made testable. The notable distinction is structural rework (i.e. refactoring) and adding features just for the sake of testing (let's say "test-driven pollution").

[the remark this answers is not in the document]
Comment (N2oB6n-SAP) [reply clean-abap/CleanABAP.md:222]: Full ack.

File: clean-abap/CleanABAP.md [hunk 2]
```diff
@@ -4589,14 +4589,16 @@ CLASS unit_tests IMPLEMENTATION.
 ENDCLASS.
 ```
 
-#### Don't change the productive code to make the code testable
+#### Don't add unit test features to productive code
 
-> [Clean ABAP](#clean-abap) > [Content](#content) > [Testing](#testing) > [Injection](#injection) > [This section](#dont-change-the-productive-code-to-make-the-code-testable)
+> [Clean ABAP](#clean-abap) > [Content](#content) > [Testing](#testing) > [Injection](#injection) > [This section](#dont add unit test features-to-productive-code)
 
+For reasons already described under [Test Seams](#use-test-seams-as-temporary-workaround), adding test features to product code should also be avoided.
```
[the remark this answers is not in the document]
Comment (N2oB6n-SAP) [reply clean-abap/CleanABAP.md:4596]: I believe we should leave even less room for ambiguity than "test features" / "test code". For example "...adding parts to the production code that are solely used for testing should also be avoided."

File: clean-abap/CleanABAP.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (N2oB6n-SAP) [reply clean-abap/CleanABAP.md:222]: I am afraid the real culprit is the german false friend compound word "Produktivcode" which managed to seep into the guide as well. For the sake of consistency we might use productive but generally speaking we really ought to switch it to something else...

**Review by N2oB6n-SAP — CHANGES_REQUESTED** [review]

File: clean-abap/CleanABAP.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (HrFlorianHoffmann) [reply clean-abap/CleanABAP.md:222]: Let's address this in a separate PR and change it from "productive" to "production" consistently.

File: clean-abap/CleanABAP.md [hunk 3]
```diff
@@ -4589,14 +4589,16 @@ CLASS unit_tests IMPLEMENTATION.
...
 ```ABAP
 " anti-pattern
 IF me->in_test_mode = abap_true.
 ```
+Note that test features for intended for the end user, e.g. simulated posting or running a report in test mode, form part of the application domain and remain a valid use case.
```
[the remark this answers is not in the document]
Comment (HrFlorianHoffmann) [reply clean-abap/CleanABAP.md:4601]: ⬆️ I like this suggestion. It's a bit clearer than the `->is_test_mode alone`, which could just as well be part of a regular piece of production code.

File: clean-abap/CleanABAP.md [hunk 4]
```diff
@@ -4589,14 +4589,16 @@ CLASS unit_tests IMPLEMENTATION.
 ENDCLASS.
 ```
 
-#### Don't change the productive code to make the code testable
+#### Don't add unit test features to productive code
```
[the remark this answers is not in the document]
Comment (N2oB6n-SAP) [reply clean-abap/CleanABAP.md]: obsolete with latest changes -> marking as resolved

**Review by N2oB6n-SAP — CHANGES_REQUESTED** [review]
I support all proposed changes (using "production" everywhere and clarifying that "test" refers to automated tests).

Only thing that is left is the tiny grammar mistake that I believe to have spotted.

**Pull request merged**